File: merge-two-binary-trees/shaun-own-dfs.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
//shaun own dfs with recur, similar to same tree or symmetric tree
//runtime 36ms, faster than 89%, mem less than 92%
class Solution {
public:
    TreeNode* mergeTrees(TreeNode* t1, TreeNode* t2) {
        if (t1 == NULL && t2 == NULL) return NULL;
        if (t1 != NULL && t2 == NULL) return t1;
        if (t1 == NULL && t2 != NULL) return t2;
        //now t1 not null and t2 not null
        t1->val += t2->val;
        t1->left = mergeTrees(t1->left, t2->left);
        t1->right= mergeTrees(t1->right, t2->right);
        return t1;
    }
};

File: diameter-of-binary-tree/shaun-own-dfs.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
/*
for each node, compute its left child tree height, right tree height
the diameter thru this node is hleft+1+hright
track the max
time  complexity is O(n^2) n is the number of nodes
space complexity is stack, height of the tree

Runtime: 1 ms, faster than 17.05% of Java online submissions for Diameter of Binary Tree.
Memory Usage: 40.8 MB, less than 5.19% of Java online submissions for Diameter of Binary Tree.

with memo
Runtime: 4 ms, faster than 14.08% of Java online submissions for Diameter of Binary Tree.
Memory Usage: 41.5 MB, less than 5.19% of Java online submissions for Diameter of Binary Tree.
*/
class Solution {
    private int diameter = 0;
    private Map<TreeNode,Integer> memo = new HashMap<>();
    public int diameterOfBinaryTree(TreeNode root) {
        helper(root);
        return diameter;
    }
    //return the height of tree, update max diameter
    private int helper(TreeNode root) {
        if (memo.containsKey(root))
            return memo.get(root);
        if (root == null) return 0;
        int left = helper(root.left);
        int right = helper(root.right);
        //diameter = Math.max(diameter, left + 1+ right);
        diameter = Math.max(diameter, left + right);
        int h = 1 + Math.max(left,right);
        memo.put(root,h);
        return h;
    }
}
